### DISCUSSION TITLE
How to set custom app settings in Microsoft.Web/sites web app?

### DISCUSSION THREAD

**Discussion by bert2** [Azure/ResourceModules#1132] [Q&A]
We're struggling with the usage of the [`Microsoft.Web/sites`](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites) module. In particular we can't find a way to provision it with our custom app settings. This is how we are trying to achieve this:

```bicep
module myWebApp './Microsoft.Web/sites/deploy.bicep' = {
    // ...
    params: {
        // ...
        siteConfig: {
            // ...
            appSettings: [
                {
                    name: 'MyCustomSetting'
                    value: 'MyCustomValue'
                }
            ]
        }
    }
}
```

The deployment works without errors, but the web app doesn't get this setting. So is this the correct way? Unfortunately we can't find any usage examples.

**Comment by bert2** [comment]
We found out that the `Microsoft.Web/sites` module seems to overwrite _anything_ that is passed as part of the `siteConfig` parameter. The way we understand it, the child module `Microsoft.Web/sites/config` (internally used by the `Microsoft.Web/sites` module) seems to create a new site configuration without incorporating the values from `siteConfig`.

Is this intended behavior?

**Comment by AlexanderSehr** [comment]
Hey @bert2, excuse the late reply. Discussions are not that commonly used in this repository. It seems the `Sites` module requires some major overhaul to properly apply the AppSettings (ref #1146). 
The properties object for the config is 'complicated' to say the least.
What you tried above does not work because the `Sites` module does not pass any unknown parameters (in this case `appSettings`) down to the confirm child-template:
```bicep
module app_appsettings 'config/deploy.bicep' = {
  name: '${uniqueString(deployment().name, location)}-Site-Config'
  params: {
    name: 'appsettings'
    appName: app.name
    storageAccountId: !empty(storageAccountId) ? storageAccountId : ''
    appInsightId: !empty(appInsightId) ? appInsightId : !empty(appInsightObject) ? appInsight.outputs.resourceId : ''
    functionsWorkerRuntime: !empty(functionsWorkerRuntime) ? functionsWorkerRuntime : ''
    functionsExtensionVersion: !empty(functionsExtensionVersion) ? functionsExtensionVersion : '~3'
    ... <- no other parameters
  }
}
```
We have discussed this issue recently and somebody should pick the issue up soon. I'll be curious to see the result.

**Reply by AlexanderSehr** [reply]
I implemented an update in #1258 that should enable you to apply your own custom app settings

**Comment by AlexanderSehr** [comment]
@bert2 Please check out the [latest module version](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.Web/sites#parameter-usage-appsettingskeyvaluepairs). It was reworked and supports CustomAppSettings (aside from a couple other features)
